### PR TITLE
allow exact duplicate params to be created without error

### PR DESCRIPTION
**Pull request by suprjinx**
…iling

┆Issue is synchronized with this [Jira Task](https://gr-oss.atlassian.net/browse/FML-50) by [Unito](https://www.unito.io)

### CODE REVIEW

File: pkg/api/mlflow/dao/repositories/param.go [hunk 1]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
 
 // CreateBatch creates []models.Param entities in batch.
 func (r ParamRepository) CreateBatch(ctx context.Context, batchSize int, params []models.Param) error {
-	if err := r.db.CreateInBatches(params, batchSize).Error; err != nil {
+	var err error
+	err = r.db.CreateInBatches(params, batchSize).Error
+	if err != nil {
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:37]: this block looks not well understandable.

File: pkg/api/mlflow/dao/repositories/param.go [hunk 2]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+		// remove duplicate rows and try again
+		params, err1 := r.removeExactMatches(ctx, params)
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:39]: this is not good to use such names like `err2` or `err2` or any other.

File: pkg/api/mlflow/dao/repositories/param.go [hunk 3]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+		if err1 != nil {
+			return eris.Wrap(err1, "error removing exact matches after: " + err.Error())
+		}
+		err = r.db.CreateInBatches(params, batchSize).Error
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:43]: it is also not a good practise to shadow upper level `err`. when we integrate linter, linter gives and error about that.

File: pkg/api/mlflow/dao/repositories/param.go [hunk 4]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
 // CreateBatch creates []models.Param entities in batch.
 func (r ParamRepository) CreateBatch(ctx context.Context, batchSize int, params []models.Param) error {
-	if err := r.db.CreateInBatches(params, batchSize).Error; err != nil {
+	var err error
+	err = r.db.CreateInBatches(params, batchSize).Error
+	if err != nil {
+		// remove duplicate rows and try again
+		params, err1 := r.removeExactMatches(ctx, params)
+		if err1 != nil {
+			return eris.Wrap(err1, "error removing exact matches after: " + err.Error())
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:41]: please do not use `err.Error()`. We have to bubble up all the errors and stack trace.

File: pkg/api/mlflow/dao/repositories/param.go [hunk 5]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
 		return eris.Wrap(err, "error creating params in batch")
 	}
 	return nil
 }
+
+// removeExactMatches will return a new slice of params which excludes exact matches
+func (r ParamRepository) removeExactMatches(ctx context.Context, params []models.Param) ([]models.Param, error) {
+	var keys, foundKeys []string
+	paramMap := map[string]models.Param{}
+	paramsToReturn := []models.Param{}
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:56]: can we please init variable new the place where we use it?

File: pkg/api/mlflow/dao/repositories/param.go [hunk 6]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+	tx := r.db.Raw(`
+           select run_uuid || '-' || key || '-' || value
+           from params
+           where run_uuid || '-' || key || '-' || value in ?`, keys).
+		Find(&foundKeys)
+	if tx.Error != nil {
+		return paramsToReturn, eris.Wrap(tx.Error, "problem selecting existing params")
+	}
+
+	for _, foundKey := range foundKeys {
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:72]: ohh, we have 3 cycles. are we 100% sure about that?

File: pkg/api/mlflow/dao/repositories/param.go [hunk 7]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+	paramMap := map[string]models.Param{}
+	paramsToReturn := []models.Param{}
+	for _, param := range params {
+		key := fmt.Sprintf("%v-%v-%v", param.RunID, param.Key, param.Value)
+		keys = append(keys, key)
+		paramMap[key] = param
+	}
+
+	tx := r.db.Raw(`
+           select run_uuid || '-' || key || '-' || value
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:64]: @jgiannuzzi @suprjinx  what about next solution -> when we create a param, lets calculate a `hash` into the own column and then let's try to use this hash to find duplicated.

File: pkg/api/mlflow/dao/repositories/param.go [hunk 8]
```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+	var keys, foundKeys []string
+	paramMap := map[string]models.Param{}
+	paramsToReturn := []models.Param{}
+	for _, param := range params {
+		key := fmt.Sprintf("%v-%v-%v", param.RunID, param.Key, param.Value)
+		keys = append(keys, key)
+		paramMap[key] = param
+	}
+
+	tx := r.db.Raw(`
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/param.go:63]: you don't need to create a var here. you can put everything into `if` statement and get `r.db.Raw().Find.Error` immediately.

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 9]
```diff
@@ -1,3 +1,190 @@
...
+	assert.Empty(s.T(), resp)
+}
+
+func (s *LogParamTestSuite) TestBatch_Ok() {
+	tests := []struct {
+		name    string
+		request *request.LogBatchRequest
+	}{
+		{
+			name: "Batch of one should succeed",
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:92]: please do not use whitespaces in test names. When test failed, all whitespaces will be removed and impossible to find failed test.

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 10]
```diff
@@ -1,3 +1,190 @@
...
+
+	runID := uuid.New().String()
+	run := &models.Run{
+		ID:             strings.ReplaceAll(runID, "-", ""),
+		ExperimentID:   *exp.ID,
+		SourceType:     "JOB",
+		LifecycleStage: models.LifecycleStageActive,
+		Status:         models.StatusRunning,
+	}
+	run, err = s.runFixtures.CreateTestRun(context.Background(), run)
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:61]: probably will be nice to create `run` in each test. In that case logic is fully isolated and not related to other tests/results.

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 11]
```diff
@@ -1,3 +1,190 @@
...
+		Value: "value1",
+	}
+	resp := map[string]any{}
+	err := s.client.DoPostRequest(
+		fmt.Sprintf("%s%s", mlflow.RunsRoutePrefix, mlflow.RunsLogParameterRoute),
+		req,
+		&resp,
+	)
+	assert.Nil(s.T(), err)
+	assert.Empty(s.T(), resp)
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:83]: probably Im not sure that we need to check that moment.

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 12]
```diff
@@ -1,3 +1,190 @@
...
+					},
+					{
+						Key:   "key2",
+						Value: "value2",
+					},
+				},
+			},
+		},
+	}
+	var err error
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:120]: what do we need this var?

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 13]
```diff
@@ -1,3 +1,190 @@
...
+		assert.Nil(s.T(), s.runFixtures.UnloadFixtures())
+	}()
+
+	var testData = []struct {
+		name    string
+		error   string
+		request *request.LogBatchRequest
+	}{
+		{
+			name:  "Missing required field",
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:146]: the same problem with naming.

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 14]
```diff
@@ -1,3 +1,190 @@
...
+		SourceType:     "JOB",
+		LifecycleStage: models.LifecycleStageActive,
+		Status:         models.StatusRunning,
+	}
+	run, err = s.runFixtures.CreateTestRun(context.Background(), run)
+	assert.Nil(s.T(), err)
+	s.run = run
+}
+
+func (s *LogParamTestSuite) Test_Ok() {
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:66]: this test is not related to `batch`, can we please move it into own test file?

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 15]
```diff
@@ -1,3 +1,190 @@
...
+		name    string
+		error   string
+		request *request.LogBatchRequest
+	}{
+		{
+			name:  "Missing required field",
+			error: "Missing value for required parameter 'run_id'",
+			request: &request.LogBatchRequest{
+				RunID: "",
+				Params: []request.ParamPartialRequest{
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:150]: probably you don't need to pass `params` here?

**Review by dsuhinin — CHANGES_REQUESTED** [review]

File: pkg/api/mlflow/dao/repositories/param.go (hunk 7 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:64]: interesting idea

File: pkg/api/mlflow/dao/repositories/param.go (hunk 6 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:72]: no sure if there's a way to avoid without some "set operations" library or the like, which also range over collections as implementation

File: pkg/api/mlflow/dao/repositories/param.go (hunk 7 above)
Comment (dsuhinin) [reply pkg/api/mlflow/dao/repositories/param.go:64]: I can see one problem here - how it should work for old records without the hash? maybe we can create a command line tool, which will migrate old data and add missing hashes?

File: pkg/api/mlflow/dao/repositories/param.go (hunk 6 above)
Comment (dsuhinin) [reply pkg/api/mlflow/dao/repositories/param.go:72]: with `hashes` we can immediately remove all duplicates and re create new ones.

File: pkg/api/mlflow/dao/repositories/param.go (hunk 1 above)
Comment (dsuhinin) [reply pkg/api/mlflow/dao/repositories/param.go:37]: let's try to use more plain constructs. now it is a bit hard to understand what is going on when you read. which error belongs to which block. maybe we can change this blocks so the could be in more plain way? without nesting?

Comment (dsuhinin) [reply pkg/api/mlflow/dao/repositories/param.go:37]: why we can't before clean? and then just create?

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 16]
```diff
@@ -1,3 +1,190 @@
...
+func (s *LogParamTestSuite) SetupTest() {
+	s.client = helpers.NewHttpClient(os.Getenv("SERVICE_BASE_URL"))
+	runFixtures, err := fixtures.NewRunFixtures(os.Getenv("DATABASE_DSN"))
+	assert.Nil(s.T(), err)
+	s.runFixtures = runFixtures
+	expFixtures, err := fixtures.NewExperimentFixtures(os.Getenv("DATABASE_DSN"))
+	assert.Nil(s.T(), err)
+	s.experimentFixtures = expFixtures
+
+	expName := uuid.New().String()
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:45]: this is not re used somewhere so you use it inline:
```
Name: uuid.New().String()
```

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 17]
```diff
@@ -1,3 +1,190 @@
...
+	exp := &models.Experiment{
+		Name:           expName,
+		LifecycleStage: models.LifecycleStageActive,
+	}
+	_, err = s.experimentFixtures.CreateTestExperiment(context.Background(), exp)
+	assert.Nil(s.T(), err)
+
+	runID := uuid.New().String()
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:53]: the same probably here.

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 13 above)
Comment (dsuhinin) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:146]: please add into the name, which required field is missing. it immediately give you all needed information when test failed.

File: pkg/api/mlflow/dao/repositories/param.go (hunk 7 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:64]: I do like this idea, especially because we could then use `ON CONFLICT(hash_column) DO NOTHING` and make things much simpler. Perhaps old records wouldn't be getting new log params, so maybe it's okay without backfill?

File: pkg/api/mlflow/dao/repositories/param.go (hunk 6 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:72]: well the issue is we have to subtract a set of keys from the hash in this implementation -- but maybe I'm not seeing it

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 17 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:53]: the compiler should inline it ;-). But I will move.

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 10 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:61]: well the Run instance isn't actually part of the test, right -- it's a fixture for Param.

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 14 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:66]: will do

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 13 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:146]: I think human readable in the test source file is probably more important than the error spit out by failure

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 17 above)
Comment (dsuhinin) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:53]: yep yep, but unfortunately we still need to support and read it =(

File: pkg/api/mlflow/dao/repositories/param.go (hunk 2 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:39]: will rename

File: pkg/api/mlflow/dao/repositories/param.go (hunk 3 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:43]: will remove var reuse

File: pkg/api/mlflow/dao/repositories/param.go (hunk 4 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:41]: will wrap

Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:41]: well at this point we have 2 errors to report and Wrap() does not seem to allow multiple error args

File: pkg/api/mlflow/dao/repositories/param.go (hunk 5 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:56]: moved

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 13 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:146]: `--- PASS: TestLogBatchTestSuite (0.03s)
    --- PASS: TestLogBatchTestSuite/Test_Error (0.03s)
        --- PASS: TestLogBatchTestSuite/Test_Error/Missing_required_field (0.00s)
        --- PASS: TestLogBatchTestSuite/Test_Error/Duplicate_keys_with_different_values_should_error (0.00s)`

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 11 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:83]: the API returns an empty body for this, seemed worth asserting

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 13 above)
Comment (dsuhinin) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:146]: convention is do not use `whitespaces` because as you can see they will be remove with `underscores` and when test failed, then you will try to fins something like `Missing_required_field` over the code but..... your test have `whitesapces`

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 15 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:150]: well to illustrate the request is otherwise well formed?

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 13 above)
Comment (dsuhinin) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:146]: so the name has to be `MissingRequiredField` + add more details about which filed is missing.

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 15 above)
Comment (dsuhinin) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:150]: but you test this case, right -> `RunID: ""`, why do you need to pass other parameters?

Comment (dsuhinin) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:150]: they are not necessary, because validation won't check them.

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 17 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:53]: breaking up code into smaller steps is often helpful for us humans :-) but I'll move this inline

File: pkg/api/mlflow/dao/repositories/param.go (hunk 4 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: now discards first error

File: pkg/api/mlflow/dao/repositories/param.go (hunk 7 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: decided to keep this in the error-only path

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 9 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go]: is this really a style recommendation in Go community? I think human-readable messages are important

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 16 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go]: done

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 15 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go]: okay will remove

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 18]
```diff
@@ -1,3 +1,102 @@
...
+		assert.Nil(s.T(), s.runFixtures.UnloadFixtures())
+	}()
+
+	// missing run_id
+	req := request.LogParamRequest{
+		Key:   "key1",
+		Value: "value1",
+	}
+	resp := map[string]any{}
+	_ = s.client.DoPostRequest(
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:95]: please check `err` here.

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 19]
```diff
@@ -1,3 +1,102 @@
...
+		fmt.Sprintf("%s%s", mlflow.RunsRoutePrefix, mlflow.RunsLogParameterRoute),
+		req,
+		&resp,
+	)
+	assert.Equal(s.T(), "INVALID_PARAMETER_VALUE", resp["error_code"])
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:100]: can we please use existing error objects to check error? - `NewInvalidParameterValueError()`

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 20]
```diff
@@ -1,3 +1,102 @@
...
+	defer func() {
+		assert.Nil(s.T(), s.runFixtures.UnloadFixtures())
+	}()
+
+	// missing run_id
+	req := request.LogParamRequest{
+		Key:   "key1",
+		Value: "value1",
+	}
+	resp := map[string]any{}
```
Comment (dsuhinin) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:94]: since you are waiting for the error, please unwrap response into `Error` response - `api.ErrorResponse{}` and then use it to check error data.

**Review by dsuhinin — CHANGES_REQUESTED** [review]
Some minor things.

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 18 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:95]: will do, but in this case the test will obviously fail if the http request is unsuccessful so it is superfluous.

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 19 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:100]: will do

File: tests/integration/golang/mlflow/run/log_parameter_test.go (hunk 20 above)
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go:94]: will do

**Review by dsuhinin — DISMISSED** [review]
looks good!

**Review by dsuhinin — DISMISSED** [review]

**Comment by suprjinx** [comment]
Moved to PR #89
